File: packages/core/src/research_engine/domain/claims.py
"""Claims, their evidence anchors, and argument-graph edges."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Any, Literal
from uuid import UUID  # noqa: TC003 - pydantic needs it at runtime

from pydantic import BaseModel, ConfigDict, Field, field_validator, model_validator


class ClaimKind(StrEnum):
    OPPOSITION = "opposition"
    MINE = "mine"
    PREMISE = "premise"
    LEXICAL = "lexical"
    ALLY = "ally"


class ClaimStatus(StrEnum):
    OPEN = "open"
    RESEARCHING = "researching"
    REBUTTED = "rebutted"
    WEAKENED = "weakened"
    UNRESOLVED = "unresolved"
    CONCEDED = "conceded"


class ClaimRelation(StrEnum):
    DEPENDS_ON = "depends_on"
    SUPPORTS = "supports"
    CONTRADICTS = "contradicts"
    REFINES = "refines"
    REBUTS = "rebuts"
    CONCEDES = "concedes"
    ENTAILS = "entails"


class AnchorRole(StrEnum):
    ASSERTS = "asserts"
    SUPPORTS = "supports"
    REBUTS = "rebuts"
    CONTEXT = "context"


class AnchorVerifyStatus(StrEnum):
    EXACT = "exact"
    NORMALIZED = "normalized"
    NEAR = "near"

CLAIM_AUDIT_ASSURANCE = (
    "Green means the implemented mechanical checks found no failure. "
    "It does not mean the argument is sound or the source has been interpreted faithfully."
)


class Claim(BaseModel):
    """One addressable proposition in ``argument.claims``."""

    id: UUID
    ref: str
    statement: str
    kind: ClaimKind
    status: ClaimStatus
    confidence: float | None = None
    steelman: str | None = None
    public_ready: bool = False
    academic_candidate: bool = False
    attributes: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime
    updated_at: datetime


class ClaimDraft(BaseModel):
    """Mutable claim fields accepted by the ledger write path."""

    model_config = ConfigDict(extra="forbid")

    ref: str
    statement: str
    kind: ClaimKind
    status: ClaimStatus = ClaimStatus.OPEN
    confidence: float | None = Field(default=None, ge=0, le=1)
    steelman: str | None = None
    public_ready: bool = False
    academic_candidate: bool = False
    attributes: dict[str, Any] = Field(default_factory=dict)

    @field_validator("ref", "statement")
    @classmethod
    def _non_empty(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("must be non-empty")
        return value


class ClaimEdge(BaseModel):
    id: UUID
    source_id: UUID
    target_id: UUID
    relation: ClaimRelation
    confidence: float | None = None
    note: str | None = None
    created_at: datetime


class ClaimEdgeDraft(BaseModel):
    """An outgoing edge named by the target's stable claim ref."""

    model_config = ConfigDict(extra="forbid")

    target_ref: str
    relation: ClaimRelation
    confidence: float | None = Field(default=None, ge=0, le=1)
    note: str | None = None

    @field_validator("target_ref")
    @classmethod
    def _target_non_empty(cls, value: str) -> str:
        value = value.strip()
        if not value:
            raise ValueError("must be non-empty")
        return value


class Anchor(BaseModel):
    id: UUID
    claim_id: UUID
    role: AnchorRole
    person_entity_id: UUID | None = None
    source_span_id: UUID
    quoted_text: str
    verify_status: AnchorVerifyStatus | None = None
    verified_at: datetime | None = None
    parser_version: str | None = None
    edition_id: UUID | None = None
    edition_key: str | None = None
    locator: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime


class AnchorDraft(BaseModel):
    """A verified anchor ready for insertion."""

    model_config = ConfigDict(extra="forbid")

    role: AnchorRole
    quoted_text: str
    source_span_id: UUID
    person_entity_id: UUID | None = None
    verify_status: AnchorVerifyStatus
    verified_at: datetime
    parser_version: str | None = None
    edition_id: UUID | None = None
    edition_key: str | None = None
    locator: dict[str, Any] = Field(default_factory=dict)

    @field_validator("quoted_text")
    @classmethod
    def _quote_non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("must be non-empty")
        return value

    @model_validator(mode="after")
    def _assertion_names_a_person(self) -> AnchorDraft:
        if self.role is AnchorRole.ASSERTS and self.person_entity_id is None:
            raise ValueError("an asserts anchor must name a person")
        return self


class AnchorInput(BaseModel):
    """Caller input before quote verification and span resolution."""

    model_config = ConfigDict(extra="forbid")

    role: AnchorRole
    quote: str
    document_id: UUID
    person: str | None = None
    edition_key: str | None = None
    locator: dict[str, Any] = Field(default_factory=dict)

    @field_validator("quote")
    @classmethod
    def _input_quote_non_empty(cls, value: str) -> str:
        if not value.strip():
            raise ValueError("must be non-empty")
        return value

    @model_validator(mode="after")
    def _assertion_input_names_a_person(self) -> AnchorInput:
        if self.role is AnchorRole.ASSERTS and not (self.person and self.person.strip()):
            raise ValueError("an asserts anchor must name a person")
        return self

class ClaimFinding(BaseModel):
    rule_id: str
    severity: Literal["error", "warning", "info"]
    claim_ref: str
    message: str
    detail: dict[str, Any] | None = None


class ClaimAuditReport(BaseModel):
    findings: list[ClaimFinding] = Field(default_factory=list)
    checked_refs: list[str] | None = None
    assurance: str




class ClaimWriteResult(BaseModel):
    claim: Claim
    edges: list[ClaimEdge] = Field(default_factory=list)
    anchors: list[Anchor] = Field(default_factory=list)
    findings: list[ClaimFinding] = Field(default_factory=list)

File: crates/marginalia-types/src/claims.rs
//! Claims, anchors, and argument-graph edges mirroring `domain/claims.py`.

use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use serde_json::{Map, Value};
use uuid::Uuid;

use crate::errors::{Error, Result};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ClaimKind {
    Opposition,
    Mine,
    Premise,
    Lexical,
    Ally,
}

#[derive(Debug, Clone, Copy, Default, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ClaimStatus {
    #[default]
    Open,
    Researching,
    Rebutted,
    Weakened,
    Unresolved,
    Conceded,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum ClaimRelation {
    DependsOn,
    Supports,
    Contradicts,
    Refines,
    Rebuts,
    Concedes,
    Entails,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AnchorRole {
    Asserts,
    Supports,
    Rebuts,
    Context,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum AnchorVerifyStatus {
    Exact,
    Normalized,
    Near,
}

/// Green means the implemented mechanical checks found no failure. It does not
/// mean the argument is sound or the source has been interpreted faithfully.
pub const CLAIM_AUDIT_ASSURANCE: &str = "Green means the implemented mechanical checks found no failure. It does not mean the argument is sound or the source has been interpreted faithfully.";

/// One addressable proposition in `argument.claims`.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Claim {
    pub id: Uuid,
    pub r#ref: String,
    pub statement: String,
    pub kind: ClaimKind,
    pub status: ClaimStatus,
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(default)]
    pub steelman: Option<String>,
    #[serde(default)]
    pub public_ready: bool,
    #[serde(default)]
    pub academic_candidate: bool,
    #[serde(default)]
    pub attributes: Map<String, Value>,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
}

/// Mutable claim fields accepted by the ledger write path.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ClaimDraft {
    pub r#ref: String,
    pub statement: String,
    pub kind: ClaimKind,
    #[serde(default)]
    pub status: ClaimStatus,
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(default)]
    pub steelman: Option<String>,
    #[serde(default)]
    pub public_ready: bool,
    #[serde(default)]
    pub academic_candidate: bool,
    #[serde(default)]
    pub attributes: Map<String, Value>,
}

impl ClaimDraft {
    /// Non-empty ref/statement plus the 0..=1 confidence bound.
    pub fn validate(&self) -> Result<()> {
        if self.r#ref.trim().is_empty() || self.statement.trim().is_empty() {
            return Err(Error::Validation("must be non-empty".to_owned()));
        }
        if let Some(c) = self.confidence {
            check_fraction(c)?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ClaimEdge {
    pub id: Uuid,
    pub source_id: Uuid,
    pub target_id: Uuid,
    pub relation: ClaimRelation,
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(default)]
    pub note: Option<String>,
    pub created_at: DateTime<Utc>,
}

/// An outgoing edge named by the target's stable claim ref.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct ClaimEdgeDraft {
    pub target_ref: String,
    pub relation: ClaimRelation,
    #[serde(default)]
    pub confidence: Option<f64>,
    #[serde(default)]
    pub note: Option<String>,
}

impl ClaimEdgeDraft {
    pub fn validate(&self) -> Result<()> {
        if self.target_ref.trim().is_empty() {
            return Err(Error::Validation("must be non-empty".to_owned()));
        }
        if let Some(c) = self.confidence {
            check_fraction(c)?;
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct Anchor {
    pub id: Uuid,
    pub claim_id: Uuid,
    pub role: AnchorRole,
    #[serde(default)]
    pub person_entity_id: Option<Uuid>,
    pub source_span_id: Uuid,
    pub quoted_text: String,
    #[serde(default)]
    pub verify_status: Option<AnchorVerifyStatus>,
    #[serde(default)]
    pub verified_at: Option<DateTime<Utc>>,
    #[serde(default)]
    pub parser_version: Option<String>,
    #[serde(default)]
    pub edition_id: Option<Uuid>,
    #[serde(default)]
    pub edition_key: Option<String>,
    #[serde(default)]
    pub locator: Map<String, Value>,
    pub created_at: DateTime<Utc>,
}

/// A verified anchor ready for insertion.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AnchorDraft {
    pub role: AnchorRole,
    pub quoted_text: String,
    pub source_span_id: Uuid,
    #[serde(default)]
    pub person_entity_id: Option<Uuid>,
    pub verify_status: AnchorVerifyStatus,
    pub verified_at: DateTime<Utc>,
    #[serde(default)]
    pub parser_version: Option<String>,
    #[serde(default)]
    pub edition_id: Option<Uuid>,
    #[serde(default)]
    pub edition_key: Option<String>,
    #[serde(default)]
    pub locator: Map<String, Value>,
}

impl AnchorDraft {
    /// Non-empty quote plus the asserts-names-a-person rule.
    pub fn validate(&self) -> Result<()> {
        if self.quoted_text.trim().is_empty() {
            return Err(Error::Validation("must be non-empty".to_owned()));
        }
        if self.role == AnchorRole::Asserts && self.person_entity_id.is_none() {
            return Err(Error::Validation(
                "an asserts anchor must name a person".to_owned(),
            ));
        }
        Ok(())
    }
}

/// Caller input before quote verification and span resolution.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(deny_unknown_fields)]
pub struct AnchorInput {
    pub role: AnchorRole,
    pub quote: String,
    pub document_id: Uuid,
    #[serde(default)]
    pub person: Option<String>,
    #[serde(default)]
    pub edition_key: Option<String>,
    #[serde(default)]
    pub locator: Map<String, Value>,
}

impl AnchorInput {
    pub fn validate(&self) -> Result<()> {
        if self.quote.trim().is_empty() {
            return Err(Error::Validation("must be non-empty".to_owned()));
        }
        if self.role == AnchorRole::Asserts
            && self.person.as_ref().is_none_or(|p| p.trim().is_empty())
        {
            return Err(Error::Validation(
                "an asserts anchor must name a person".to_owned(),
            ));
        }
        Ok(())
    }
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ClaimFinding {
    pub rule_id: String,
    pub severity: Severity,
    pub claim_ref: String,
    pub message: String,
    #[serde(default)]
    pub detail: Option<Map<String, Value>>,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Severity {
    Error,
    Warning,
    Info,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ClaimAuditReport {
    #[serde(default)]
    pub findings: Vec<ClaimFinding>,
    #[serde(default)]
    pub checked_refs: Option<Vec<String>>,
    pub assurance: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
pub struct ClaimWriteResult {
    pub claim: Claim,
    #[serde(default)]
    pub edges: Vec<ClaimEdge>,
    #[serde(default)]
    pub anchors: Vec<Anchor>,
    #[serde(default)]
    pub findings: Vec<ClaimFinding>,
}

fn check_fraction(value: f64) -> Result<()> {
    if !(0.0..=1.0).contains(&value) {
        return Err(Error::Validation(
            "confidence must be between 0 and 1".to_owned(),
        ));
    }
    Ok(())
}
